src/typechecker.erl

```diff
@@ -20,6 +20,36 @@
 
 -type type() :: erl_parse:abstract_type().
 
+-type type_error() ::
+    #{error_type := {undef, call | record | user_type | remote_type | variable}
+                    | unexported_remote_type | var | char | atom
+                    | string | int | float | improper_tail | cons_pat
+                    | cons | nil | argument_length_mismatch | call_args
+                    | {call, args | arity | context} | mfa
+                    | return | expected_fun_type | op_type_too_precise
+                    | {arith_op, arguments | expression}
+                    | {int_op, arguments | expression}
+                    | non_number_argument_to_plus | non_number_argument_to_minus
+                    | non_boolean_argument_to_not
+                    | unary_op | non_integer_argument_to_bnot
+                    | logic_op | {rel_op, arguments | expression}
+                    | {list_op, arguments | expression}
+                    | operator_pattern | pattern | tuple
```

Comment:
These are way too many IMO. Some of these (var, char, atom, string, int, float, cons, nil, ......) can either appear as a pattern or as an expression. Exactly which pattern or expression it is is already encoded in the expression field, but we can't see if it appears as a pattern or as an expression because these are syntactically the same. So, what we really need to distinguish between here is `expression | pattern`. Then as a helper for pretty-printing we can have simple function for describing an expression, e.g.

```Erlang
describe_expr(Expr) ->
    case element(1, Expr) of
        lc    -> "list comprehension";
        'case'-> "case expression";
        var   -> "variable " ++ erl_pp:expr(Expr);
        Other -> atom_to_list(Other)
    end.
```

Possibly, other error_types can be merged too, but I haven't thought much about which and how.